examples/generative/ddpm.py

```diff
@@ -0,0 +1,682 @@
+widely used for implementing diffusion models but with some slight modifications:
+
+1. The network accepts two inputs: Image and time step
+2. Self-attention between the convolution blocks once we reach a specific resolution
+(16x16 in the paper)
+3. Group Normalization instead of weight normalization
+
+We will implement most of the things as used in the original paper. We will use `swish`
+activation function throughout the network. We will use variance scaling kernel
+initializer. 
+
+The only difference here is the number of groups used for GroupNormalization layer.
+For the flowers dataset, I found that a value of `groups=8` produces better results
+compared to the default value of `groups=32`. Dropout is optional and should be
+used where chances of over fitting is high. In the paper, the authors used dropout
+only when training on CIFAR10.
+"""
+
+
+# Activation function to use
+act_fn = keras.activations.swish
+
+
+# Kernel initializer to use
+def kernel_init(scale):
+    if scale == 0.0:
+        scale = 1e-10
+    return keras.initializers.VarianceScaling(
+        scale, mode="fan_avg", distribution="uniform"
+    )
+
+
+def AttentionBlock(units, groups=8):
+    def apply(inputs):
+        batch_size = tf.shape(inputs)[0]
+        height = tf.shape(inputs)[1]
+        width = tf.shape(inputs)[2]
+        scale = tf.cast(units, tf.float32) ** (-0.5)
+
+        inputs = layers.GroupNormalization(groups=groups)(inputs)
+        q = layers.Dense(units, kernel_initializer=kernel_init(1.0))(inputs)
+        k = layers.Dense(units, kernel_initializer=kernel_init(1.0))(inputs)
+        v = layers.Dense(units, kernel_initializer=kernel_init(1.0))(inputs)
+
+        attn_score = tf.einsum("bhwc, bHWc->bhwHW", q, k) * scale
+        attn_score = tf.reshape(attn_score, [batch_size, height, width, height * width])
+
+        attn_score = tf.nn.softmax(attn_score, -1)
+        attn_score = tf.reshape(attn_score, [batch_size, height, width, height, width])
+
+        proj = tf.einsum("bhwHW,bHWc->bhwc", attn_score, v)
+        proj = layers.Dense(units, kernel_initializer=kernel_init(0.0))(inputs)
+        return inputs + proj
+
+    return apply
+
+
+def TimeEmbedding(inputs, dim=32):
+    """Transformer sinusoidal position embedding."""
+    half_dim = dim // 2
+    emb = math.log(10000) / (half_dim - 1)
+    emb = tf.exp(tf.range(half_dim, dtype=tf.float32) * -emb)
+
+    inputs = tf.cast(inputs, dtype=tf.float32)
+    emb = inputs[:, None] * emb[None, :]
+    emb = tf.concat([tf.sin(emb), tf.cos(emb)], axis=-1)
+    return emb
+
+
+def ResidualBlock(width, groups=8):
+    def apply(inputs):
+        x, t = inputs
+        input_width = x.shape[3]
+
+        if input_width == width:
+            residual = x
+        else:
+            residual = layers.Conv2D(
+                width, kernel_size=1, kernel_initializer=kernel_init(1.0)
+            )(x)
+
+        temb = act_fn(t)
```

Comment:
Shouldn't this be an argument?

Reply:
Are you asking about `t`? The inputs to residual block is a list i.e. [x, t]